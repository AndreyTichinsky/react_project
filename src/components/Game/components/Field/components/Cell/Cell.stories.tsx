import React from "react";
import { action } from "@storybook/addon-actions";
import { withKnobs, number, boolean } from "@storybook/addon-knobs";

import { Cell } from "./Cell";

export default {
  title: "Cell",
  decorators: [withKnobs],
};

export const emptyCell = () => {
  return (
    <Cell
      isAlive={boolean("isAlive", false)}
      x={number("x", 0)}
      y={number("y", 0)}
      cellSize={number("cellSize", 25)}
      onClick={action("Cell clicked")}
      animationSpeed={number("animationSpeed", 500)}
    />
  );
};
