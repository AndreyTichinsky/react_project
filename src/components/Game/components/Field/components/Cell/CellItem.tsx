import styled from "@emotion/styled";
import { css, keyframes } from "@emotion/core";

const BaseCell = css`
  position: absolute;
  border: 1px solid #000;
  box-sizing: border-box;
  transition: background-color 0.3s ease;
`;

const aliveCell = css`
  background-color: #000000;
`;

const deathCell = css`
  background-color: #ffffff;
`;

interface Props {
  isAlive: boolean;
  cellSize: number;
  x: number;
  y: number;
  decideWhichAnimation: boolean;
}

export const CellItem = styled.div`
  width: ${({ cellSize }: Props) => cellSize}px;
  height: ${({ cellSize }: Props) => cellSize}px;
  top: ${({ cellSize, y }: Props) => cellSize * y}px;
  left: ${({ cellSize, x }: Props) => cellSize * x}px;
  ${BaseCell};
  ${({ decideWhichAnimation }: Props) => {
    return decideWhichAnimation ? aliveCell : deathCell;
  }};
`;
