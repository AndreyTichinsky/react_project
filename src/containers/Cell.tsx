import { connect } from "react-redux";
import { Cell } from "@/components/Game/components/Field/components/Cell/Cell";
import { State } from "@/redux/reducers";
import { makeConvertSpeedToNumber, makeGetAlive } from "@/redux/selectors";

const makeMapStateToProps = () => {
  const convertSpeedToNumber = makeConvertSpeedToNumber();
  const getAlive = makeGetAlive();
  const mapStateToProps = (state: State, props: any) => {
    return {
      cellSize: state.cellSize,
      isAlive: getAlive(state, props),
      animationSpeed: convertSpeedToNumber(state),
    };
  };
  return mapStateToProps;
};

export default connect(makeMapStateToProps)(Cell);
