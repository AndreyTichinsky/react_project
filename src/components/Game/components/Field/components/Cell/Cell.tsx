import React, { FC, PureComponent } from "react";
import type { CellProps } from "types/field";
import { CellItem } from "./CellItem";

export class Cell extends React.PureComponent<CellProps> {
  handleClick = () => {
    this.props.onClick(this.props.x, this.props.y);
  };
  render() {
    return (
      <CellItem
        className={`cell cell_${this.props.y}_${this.props.x}`}
        isAlive={this.props.isAlive}
        x={this.props.x}
        y={this.props.y}
        cellSize={this.props.cellSize}
        onClick={this.handleClick}
        animationSpeed={this.props.animationSpeed}
      />
    );
  }
}
