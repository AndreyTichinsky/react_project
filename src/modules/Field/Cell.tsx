import React, { FC, useCallback } from "react";
import { CellItem } from "./CellItem";
import { connect } from "react-redux";
import { AppDispatch } from "@/redux/store";
import { State } from "@/redux/reducers";
import { setCellState } from "@/redux/actions";
import { makeConvertSpeedToNumber, makeGetAlive } from "@/redux/selectors";

export interface CellProps {
  isAlive: boolean;
  x: number;
  y: number;
  cellSize: number;
  animationSpeed: number;
}

const makeMapStateToProps = () => {
  const convertSpeedToNumber = makeConvertSpeedToNumber();
  const getAlive = makeGetAlive();
  const mapStateToProps = (state: State, props: any) => {
    return {
      cellSize: state.cellSize,
      isAlive: getAlive(state, props),
      animationSpeed: convertSpeedToNumber(state),
    };
  };
  return mapStateToProps;
};

export const CellComponent: FC<CellProps & { dispatch: AppDispatch }> = (
  props
) => {
  const onClick = useCallback(() => {
    props.dispatch(setCellState({ x: props.x, y: props.y }));
  }, []);
  return (
    <CellItem
      className={`cell cell_${props.y}_${props.x}`}
      isAlive={props.isAlive}
      x={props.x}
      y={props.y}
      cellSize={props.cellSize}
      onClick={onClick}
      animationSpeed={props.animationSpeed}
    />
  );
};

export const Cell = connect(makeMapStateToProps)(CellComponent);
