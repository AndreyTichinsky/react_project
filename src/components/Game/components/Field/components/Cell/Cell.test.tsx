import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";

import { Cell } from "./Cell";

describe("Cell", () => {
  it("empty cell test", () => {
    expect(
      renderer
        .create(
          <Cell isAlive={false} x={0} y={0} cellSize={50} onClick={jest.fn()} />
        )
        .toJSON()
    ).toMatchSnapshot();
  });
  it("filled cell test", () => {
    expect(
      renderer
        .create(
          <Cell isAlive={true} x={0} y={0} cellSize={50} onClick={jest.fn()} />
        )
        .toJSON()
    ).toMatchSnapshot();
  });
  it("check prop decideWhichAnimation value 1", () => {
    const wrapper = mount(
      <Cell isAlive={false} x={0} y={0} cellSize={50} onClick={jest.fn()} />
    );
    expect(wrapper.state("decideWhichAnimation")).toEqual(false);
  });
  it("check prop decideWhichAnimation value 2", () => {
    const wrapper = mount(
      <Cell isAlive={false} x={0} y={0} cellSize={50} onClick={jest.fn()} />
    );
    wrapper.setProps({ isAlive: true });
    expect(wrapper.state("decideWhichAnimation")).toEqual(true);
  });
});
