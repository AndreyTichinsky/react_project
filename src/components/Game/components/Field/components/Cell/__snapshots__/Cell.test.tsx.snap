// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell empty cell test 1`] = `
.emotion-0 {
  width: 50px;
  height: 50px;
  top: 0px;
  left: 0px;
  position: absolute;
  border: 1px solid #000;
  box-sizing: border-box;
  -webkit-transition: background-color 0.3s ease;
  transition: background-color 0.3s ease;
  background-color: #ffffff;
}

<div
  className="cell cell_0_0 emotion-0"
  onClick={[Function]}
  x={0}
  y={0}
/>
`;

exports[`Cell filled cell test 1`] = `
.emotion-0 {
  width: 50px;
  height: 50px;
  top: 0px;
  left: 0px;
  position: absolute;
  border: 1px solid #000;
  box-sizing: border-box;
  -webkit-transition: background-color 0.3s ease;
  transition: background-color 0.3s ease;
  background-color: #000000;
}

<div
  className="cell cell_0_0 emotion-0"
  onClick={[Function]}
  x={0}
  y={0}
/>
`;
